### PR TITLE
Rabbit queue filter

### PR DESCRIPTION
**Pull request by e0jopeka**

### CODE REVIEW

File: EventListener/DiscoverUrlListener.php [hunk 1]
```diff
@@ -28,16 +29,31 @@ class DiscoverUrlListener
     private $eventDispatcher;
 
     /**
+     * @var Client
+     */
+    private $redis;
+
+    /**
+     * @var
```
Comment (lkalinka) [inline EventListener/DiscoverUrlListener.php]: add missing type to this docblock

File: EventListener/DiscoverUrlListener.php [hunk 2]
```diff
@@ -28,16 +29,31 @@ class DiscoverUrlListener
     private $eventDispatcher;
 
     /**
+     * @var Client
```
Comment (lkalinka) [inline EventListener/DiscoverUrlListener.php]: can you add full namespace there ?

File: EventListener/DiscoverUrlListener.php [hunk 3]
```diff
@@ -78,11 +91,39 @@ public function onDiscoverUrl(Event $event)
...
-                }
+                $this->queue->publishJob($job);
             }
         }
     }
 
+    /**
+     * Check if uri was added to queue before.
+     *
+     * @param $uri
```
Comment (lkalinka) [inline EventListener/DiscoverUrlListener.php]: missing type in both parameters

File: Indexer.php [hunk 4]
```diff
@@ -73,13 +73,91 @@ public function isUrlIndexedandNotExpired($uri, array $metadata = [])
...
-
         $result = $this->client->select($query);
 
         return ($result->getNumFound() > 0);
     }
 
     /**
+     * Returns unique hashes for uris from one page.
+     *
+     * @param $uris
```
Comment (lkalinka) [inline Indexer.php]: missing type with full namespace there

File: Indexer.php [hunk 5]
```diff
@@ -73,13 +73,91 @@ public function isUrlIndexedandNotExpired($uri, array $metadata = [])
...
+            $uris = $this->getUnstoredOrExpiredUris($uris, $storedIds);
+        }
+
+        return $uris;
+    }
+
+    /**
+     * Returns unique urls to be crawl from one page.
+     *
+     * @param $uris
```
Comment (lkalinka) [inline Indexer.php]: missing type in docblock

File: Tests/IndexerTest.php [hunk 6]
```diff
@@ -401,4 +404,210 @@ public function ifUrlIsNotIndexedReturnTrue()
...
+
+    public function urisNotUniqueDataProvider()
+    {
+        $url1 = new Uri('http://www.simgroep.nl/internet/medewerkers_41499/');
+        $url2 = new Uri('http://www.simgroep.nl/internet/medewerkers_41499/');
+        $url3 = new Uri('http://www.simgroep.nl/internet/vacatures_41521/');
+        $url4 = new Uri('http://www.simgroep.nl/internet/vacatures_41521/');
+
+        return [
+            [[$url1, $url2, $url3, $url4]]
```
Comment (websid) [inline Tests/IndexerTest.php]: do we really need this variables $url1...4 ? Can you return them in array without using temp variables?

File: Indexer.php [hunk 7]
```diff
@@ -322,4 +401,21 @@ protected function addDocuments(Query $updateQuery, array $documents)
...
+     *
+     * @return string
+     */
+    public function getHashSolarId($uri)
+    {
+        if (is_string($uri)) {
+            return sha1(strtolower(UrlCheck::fixUrl($uri)));
+        }
+
+        return sha1(strtolower(UrlCheck::fixUrl($uri->toString())));
```
Comment (lkalinka) [inline Indexer.php]: can you extract this sha1(strtolower(UrlCheck:: ( to seprate method

File: Indexer.php [hunk 8]
```diff
@@ -322,4 +401,21 @@ protected function addDocuments(Query $updateQuery, array $documents)
 
         $this->client->update($updateQuery);
     }
+
+
+    /**
+     * Get unique solar index document id
+     *
+     * @param $uri
```
Comment (lkalinka) [inline Indexer.php]: missing docblock probably 2 thisngs will enter there

File: Tests/IndexerTest.php [hunk 9]
```diff
@@ -401,4 +404,210 @@ public function ifUrlIsNotIndexedReturnTrue()
...
+        $result = $indexer->getUnstoredOrExpiredUris($urlUnstored, $storedIds);
+        $this->assertArrayHasKey(sha1(strtolower(UrlCheck::fixUrl($urlUnstored[0]))), $result);
+
+        $result = $indexer->getUnstoredOrExpiredUris($uris, $storedIds);
+        $this->assertEmpty($result);
+    }
+
+    public function urisNotUniqueDataProvider()
+    {
+        $url1 = new Uri('http://www.simgroep.nl/internet/medewerkers_41499/');
```
Comment (websid) [inline Tests/IndexerTest.php]: do we really need this variables $url1...4 ? Can you return them in array without using temp variables?

File: Tests/IndexerTest.php [hunk 10]
```diff
@@ -401,4 +404,210 @@ public function ifUrlIsNotIndexedReturnTrue()
...
+        $this->assertEquals($indexer->getHashSolarId($uriObj), $hashIdUri);
+        $this->assertEquals($hashIdUrl, $hashIdUri);
+    }
+
+    public function storedAndNotExpiredDataProvider()
+    {
+        $url1 = 'http://www.simgroep.nl/internet/medewerkers_41499/';
+        $url2 = 'http://www.simgroep.nl/internet/nieuws-uit-de-branche_41509/';
+        $url3 = 'http://www.simgroep.nl/internet/portfolio_41515/search/';
+        $url4 = 'http://www.simgroep.nl/internet/vacatures_41521/';
```
Comment (lkalinka) [inline Tests/IndexerTest.php]: could one of them dont have a / at end to see if UrlCheck working properly and will not be a problem with slashes at end

File: Indexer.php [hunk 11]
```diff
@@ -73,13 +73,91 @@ public function isUrlIndexedandNotExpired($uri, array $metadata = [])
...
     }
 
     /**
+     * Returns unique hashes for uris from one page.
+     *
+     * @param $uris
+     *
+     * @return array
+     */
+    public function getUniqueHashIds($uris)
```
Comment (websid) [inline Indexer.php]: That function can be refactored. You add $HashId in array key end later you retrieving keys from arrays. Why you just do not use syntax like $hashes[] = $hashId; return $hashes;

File: Tests/IndexerTest.php [hunk 12]
```diff
@@ -401,4 +404,210 @@ public function ifUrlIsNotIndexedReturnTrue()
...
+
+        $result = $indexer->getUnstoredOrExpiredUris($uris, $storedIds);
+        $this->assertEmpty($result);
+    }
+
+    public function urisNotUniqueDataProvider()
+    {
+        $url1 = new Uri('http://www.simgroep.nl/internet/medewerkers_41499/');
+        $url2 = new Uri('http://www.simgroep.nl/internet/medewerkers_41499/');
+        $url3 = new Uri('http://www.simgroep.nl/internet/vacatures_41521/');
```
Comment (lkalinka) [inline Tests/IndexerTest.php]: add also there something without slash at end

File: UrlCheck.php [hunk 13]
```diff
@@ -116,4 +119,18 @@ public static function fixUrl ($url) {
 
         return $url;
     }
+
+    /**
+     * Returns uri without hashed suffixes.
+     * @param $uri
```
Comment (lkalinka) [inline UrlCheck.php]: type missing

**Comment by lkalinka** [comment]
+1

**Pull request merged**